### PR TITLE
New model for CommandTrigger Tab-Completion.

### PR DESCRIPTION
**Pull request by rkdrns4747**
<h1> New Command Trigger Tab-Completion Model Suggestion  </h1>

<h2> Command Usage </h2>

**NOT YET IMPLEMENTED**

<h2> JSON </h2>

Tested and Fully functional on
- [x] Bukkit-latest
- [ ] Bukkit-legacy
- [ ] Sponge

Example JSON:
```json
{
  "tabs": [
    {
		"index":0,
		"candidates": "give,get,astro"
    },
	{
		"index":1,
		"candidates": "$playerlist",
		"conditions": [
			{
				"index": 0,
				"regex": "give|get"
			}
		]
	},
	{
		"index":1,
		"hint": "<amount>",
		"conditions": [
			{
				"index": 0,
				"regex": "astro"
			}
		]
	},
	{
		"index":2,
		"hint": "<amount>",
		"conditions": [
			{
				"index": 0,
				"regex": "give|get"
			}
		]
	},
	{
		"index":2,
		"hint":"<another>",
		"candidates": "aaaa,bbbb,cccc",
		"conditions": [
			{
				"index": 0,
				"regex": "astro"
			},
			{
				"index": 1,
				"regex": "100"
			}		
		]
	}	
  ],
  "aliases": [],
  "permissions": []
}
```

<h3> INDEX </h3>
As you can see, you can now specify INDEX of each tab-completer. Still, if "index" is missed, the order(index) of the object in list would be the "index" value like before. Also, you may add more than one tab-completer for one index, so that you can implement different hint/candidate based on "conditions". 

---
<h3> CONDITIONS </h3>
A new feature.  You can now test specific String args[index]  with regex.
<h4> INDEX in CONDITION</h4>
The index of arguments that should be tested.

<h4> REGEX in CONDITION</h4>
The regex string that should be used for test.

So, in conclusion, the CONDTION acts like
```java
matches(args[INDEX], REGEX)
```

And also, as you can see on example , you can add multiple CONDITION at one tab-completer, so that checking multiple arguments would be available (they treated as AND expression).

### CODE REVIEW

**Comment by wysohn** [comment]
I remembered that you were working on the command system

I have some intersections where I'm working on right now, so let me know if you have any concerns

https://github.com/TriggerReactor/TriggerReactor/commits/development_separate_listeners

**Comment by rkdrns4747** [comment]
@wysohn 
> I remembered that you were working on the command system
> 
> I have some intersections where I'm working on right now, so let me know if you have any concerns
> 
> https://github.com/TriggerReactor/TriggerReactor/commits/development_separate_listeners
Hmm yeah, looks like there would be some conflict within two pulls. I'll look into it.

**Comment by wysohn** [comment]
https://github.com/TriggerReactor/TriggerReactor/blob/development_separate_listeners/docs/command_registration.jpg

fyi, this is the new registration structure so that we can decouple the classes

**Comment by rkdrns4747** [comment]
> https://github.com/TriggerReactor/TriggerReactor/blob/development_separate_listeners/docs/command_registration.jpg
> 
> fyi, this is the new registration structure so that we can decouple the classes

@wysohn I just checked out the changes. As we expected, there's 'lot of conflicts, and looks like I have to rebase this branch to `development_separate_listeners` and re-implement this logic. I already have implemented code so not a big deal, Let me create one more branch for this PR #491 .

**Comment by wysohn** [comment]
> > https://github.com/TriggerReactor/TriggerReactor/blob/development_separate_listeners/docs/command_registration.jpg
> > fyi, this is the new registration structure so that we can decouple the classes
> 
> @wysohn I just checked out the changes. As we expected, there's 'lot of conflicts, and looks like I have to rebase this branch to `development_separate_listeners` and re-implement this logic. I already have implemented code so not a big deal, Let me create one more branch for this PR #491 .

Thanks a lot!

Sorry that I should've mentioned it before though.

It's kind of you to do that but If it's too much work, I can rebase on my side too.

**Comment by rkdrns4747** [comment]
> > > https://github.com/TriggerReactor/TriggerReactor/blob/development_separate_listeners/docs/command_registration.jpg
> > > fyi, this is the new registration structure so that we can decouple the classes
> > 
> > @wysohn I just checked out the changes. As we expected, there's 'lot of conflicts, and looks like I have to rebase this branch to `development_separate_listeners` and re-implement this logic. I already have implemented code so not a big deal, Let me create one more branch for this PR #491 .
> 
> Thanks a lot!
> 
> Sorry that I should've mentioned it before though.
> 
> It's kind of you to do that but If it's too much work, I can rebase on my side too.

Nah, It's alright. I'll work on it.

**Comment by rkdrns4747** [comment]
Origin of PR #501 , this PR would be closed due to rebasing code to PR #500
